### PR TITLE
--Articulated Object Attributes/Managers support

### PR DESCRIPTION
**Pull request by jturner65**
## Motivation and Context
This PR adds support for articulated object configurations as Attributes, and their management via an AttributeManager, in the same way that Object Attributes, Stage Attributes, etc, are managed.

We allowed for articulated object json files already, to describe habitat-specific fields as well as certain overrides, but they were not integrated into the Metadata system and so were of limited usefulness. 

~~This is currently a WIP.~~ This PR is now ready to go!

 The handling of configs, either loaded directly if specified with .json extension, or built on the fly if specified filename is the source URDF file, currently works, and these configs are being used explicitly during AO instantiation to reference the URDF file and build the articulated objects.
 

## How Has This Been Tested
All existing c++ and python tests pass.

More tests will be added to test the AO Attributes and Managers' functionality directly.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/bindings/AttributesBindings.cpp [hunk 1]
```diff
@@ -187,6 +189,34 @@ void initAttributesBindings(py::module& m) {
...
+      .def_property(
+          "shader_type", &ArticulatedObjectAttributes::getShaderType,
+          &ArticulatedObjectAttributes::setShaderType,
+          R"(The shader type [0=material, 1=flat, 2=phong, 3=pbr] to use for this construction.
+          Currently Articulated Objects only support Flat/Phong shading.)")
+      .def_property(
+          "render_mode", &ArticulatedObjectAttributes::getRenderMode,
+          &ArticulatedObjectAttributes::setRenderMode,
+          R"(Whether we should render using the articulated object, its skin,
+          primitives representing each link, both or none.)");
```
[the remark this answers is not in the document]
Comment (jturner65) [reply src/esp/bindings/AttributesBindings.cpp:224]: After our dicussion from yesterday, I'm going to keep this field around for now, with the understanding that its format or function might change in the future.

File: src/esp/metadata/attributes/AttributesEnumMaps.h [hunk 2]
```diff
@@ -17,6 +17,45 @@ enum class MotionType;
...
+   */
+  Default,
+  /**
+   * Render the articulated object using its skin.
+   */
+  Skin,
+  /**
+   * Render the articulated object using primitives to respresent each link.
+   */
+  Primitives,
```
Comment (aclegg3) [inline src/esp/metadata/attributes/AttributesEnumMaps.h:43]: Let's call this `LinkVisuals` to be explicit that this means rendering the `<visual>` tag in the `<link>` which is typically a `mesh` type.

**Comment by jturner65** [comment]
There's a couple linking errors in a few of the tests that I have to chase down, but functionally this is complete. Yay!

File: src/esp/bindings/PhysicsWrapperManagerBindings.cpp [hunk 3]
```diff
@@ -276,14 +276,38 @@ void initPhysicsWrapperManagerBindings(pybind11::module& m) {
...
+          Optionally force the articulated object's model to be reloaded from disk and assign its initial
+          LightSetup key. Returns a reference to the created object.)")
+      .def(
+          "add_articulated_object_from_urdf",
+#ifdef ESP_BUILD_WITH_BULLET
+          &ArticulatedObjectManager::addBulletArticulatedObjectByID,
+#else
+          &ArticulatedObjectManager::addArticulatedObjectByID,
+#endif
+          "filepath"_a, "force_reload"_a = false,
```
Comment (aclegg3) [inline src/esp/bindings/PhysicsWrapperManagerBindings.cpp:298]: ?
```suggestion
          "ao_lib_id"_a, "force_reload"_a = false,
```

File: src/esp/physics/bullet/BulletPhysicsManager.cpp [hunk 4]
```diff
@@ -204,8 +188,8 @@ int BulletPhysicsManager::addArticulatedObjectFromURDF(
 
   // render visual shapes if either no skinned mesh is present or if the debug
   // flag is enabled
-  bool renderVisualShapes =
-      !renderAssetPath || model->getDebugRenderPrimitives();
+  bool renderVisualShapes = !renderAssetPath || (*renderAssetPath == "") ||
+                            model->getDebugRenderPrimitives();
```
Comment (aclegg3) [inline src/esp/physics/bullet/BulletPhysicsManager.cpp:192]: I know it didn't originate with this PR, but can we refactor away this "primitives" an "debug" verbiage throughout the API?
'\<visual\>' elements are not necessarily primitives or exclusively for debugging. I think this confuses things.
Something like "visual elements" or "link visual shapes" would be much better.

File: src/tests/AttributesManagersTest.cpp [hunk 5]
```diff
@@ -527,7 +530,7 @@ void AttributesManagersTest::testAssetAttributesModRegRemove(
   // verify new handle is in template library
   CORRADE_INTERNAL_ASSERT(
       assetAttributesManager_->getObjectLibHasHandle(newHandle));
-  // verify old template is still present as well
+  // verify old template is still present as wellPbrShaderAttributesManager
```
Comment (aclegg3) [inline src/tests/AttributesManagersTest.cpp:533]: Typo?

**Review by aclegg3 — COMMENTED** [review]
Thanks! Looks good overall. Noted some minor typos and small refactor requests.

File: src/esp/metadata/attributes/ArticulatedObjectAttributes.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  setURDFPath(handle);
+  setRenderAssetHandle("");
+  setSemanticId(0);
+  // Set the default base type to be free joint
+  setBaseType(getAOBaseTypeName(ArticulatedObjectBaseType::Free));
+  // Set the default source for the interia calculation to be computed
+  setInertiaSource(
+      getAOInertiaSourceName(ArticulatedObjectInertiaSource::Computed));
+  // Set the default link order to use as the order specified in the URDF
+  setLinkOrder(getAOLinkOrderName(ArticulatedObjectLinkOrder::URDFOrder));
```
[the remark this answers is not in the document]
Comment (jturner65) [reply src/esp/metadata/attributes/ArticulatedObjectAttributes.cpp:23]: Oh, cool. I wasn't aware.

File: src/esp/metadata/managers/SceneInstanceAttributesManager.cpp [hunk 7]
```diff
@@ -292,10 +292,39 @@ SceneInstanceAttributesManager::createAOInstanceAttributesFromJSON(
 
   // only used for articulated objects
   // fixed base
-  io::jsonIntoSetter<bool>(jCell, "fixed_base",
-                           [instanceAttrs](bool fixed_base) {
-                             instanceAttrs->setFixedBase(fixed_base);
-                           });
+  // TODO remove this setter - only used enum-backed values in scene instances
```
[the remark this answers is not in the document]
Comment (jturner65) [reply src/esp/metadata/managers/SceneInstanceAttributesManager.cpp:295]: Yeah.  Not going to remove the load support for SceneInstance fixed_base boolean right now, just adding the support for the various enums - when we can fix the scene instances in the various datasets, then we can revisit this.  Any scene instance value for 'fixed_base' will override whatever is in a loaded config, at least until we make these dataset changes.

File: src/esp/physics/PhysicsManager.cpp [hunk 8]
```diff
@@ -360,26 +437,109 @@ int PhysicsManager::addArticulatedObjectInstance(
...
+    const std::string& lightSetup) {
+  if (simulator_ != nullptr) {
+    // aquire context if available
+    simulator_->getRenderGLContext();
+    auto& drawables = simulator_->getDrawableGroup();
+    return addArticulatedObjectFromURDF(
+        filepath, &drawables, fixedBase, globalScale, massScale, forceReload,
+        maintainLinkOrder, intertiaFromURDF, lightSetup);
+  }
+  // TODO Support non-existent simulator?
```
[the remark this answers is not in the document]
Comment (jturner65) [reply src/esp/physics/PhysicsManager.cpp:474]: Heh that's right.  Not sure why we even perform that check anymore.  Ah well, something for BE week :D .

**Review by aclegg3 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**